### PR TITLE
CSB-27: Add Translation Pipeline

### PR DESCRIPTION
**Pull request by jan-Kulpas**

### CODE REVIEW

File: model_utilities/__main__.py [hunk 1]
```diff
@@ -24,30 +24,38 @@ def train_model(config: dict, logger: Logger) -> None:
...
+    # Default model commented out to make comparing results easy
+    # model = AutoModelForSeq2SeqLM.from_pretrained("facebook/nllb-200-distilled-600M")
+    # tokenizer = NllbTokenizer.from_pretrained("facebook/nllb-200-distilled-600M")
+
+    model = AutoModelForSeq2SeqLM.from_pretrained(output_model_name)
     tokenizer = NllbTokenizer.from_pretrained(output_model_name)
     message = "Wsiądźmy do tego autobusu"
-    translated_message = Translator(logger, pretrained_model, tokenizer).translate(message)
+
+    translated_message = Translator(logger, model, tokenizer).translate(message, "pol_Latn", "por_Latn")
```
Comment (filip-szweda) [inline model_utilities/__main__.py:35]:
```suggestion
    translated_message = Translator(logger, model, tokenizer).translate(message, "pol_Latn", "csb_Latn")
```

File: model_utilities/__main__.py [hunk 2]
```diff
@@ -24,30 +24,38 @@ def train_model(config: dict, logger: Logger) -> None:
...
     tokenizer = NllbTokenizer.from_pretrained(output_model_name)
     eval_data = data_loader.load_data(config["DATA"]["evaluation_data_file"])
 
-    source_language = eval_data[eval_data.columns[0]]
-    target_language = eval_data[eval_data.columns[1]]
+    source_data = eval_data[eval_data.columns[0]]
+    target_data = eval_data[eval_data.columns[1]]
 
-    evaluator = ModelEvaluator(logger)
+    evaluator: ModelEvaluator = ModelEvaluator(logger, model, tokenizer)
```
Comment (filip-szweda) [inline model_utilities/__main__.py:50]:
```suggestion
    evaluator = ModelEvaluator(logger, model, tokenizer)
```

File: model_utilities/__main__.py [hunk 3]
```diff
@@ -24,30 +24,38 @@ def train_model(config: dict, logger: Logger) -> None:
...
 
-    result = evaluator.evaluate(pretrained_model, tokenizer, sentences=source_language, references=target_language)
-    print(f"BLEU Score ({source_language.name} -> {target_language.name}): {result.score}")
+    bleu, chrfpp = evaluator.evaluate(sentences=source_data, references=target_data)
```
Comment (filip-szweda) [inline model_utilities/__main__.py:52]:
```suggestion
    bleu, chr_fpp = evaluator.evaluate(sentences=source_data, references=target_data)
```

File: model_utilities/evaluate/evaluator.py [hunk 4]
```diff
@@ -10,28 +10,21 @@
 class ModelEvaluator:
     __logger: Logger
 
-    def __init__(self, logger) -> None:
+    def __init__(self, logger, model, tokenizer) -> None:
```
Comment (filip-szweda) [inline model_utilities/evaluate/evaluator.py:13]: Type hints.

File: model_utilities/evaluate/evaluator.py [hunk 5]
```diff
@@ -10,28 +10,21 @@
...
+    def evaluate(self, sentences: pd.Series, references: pd.Series) -> tuple[sacrebleu.metrics.BLEUScore, sacrebleu.metrics.CHRFScore]:
         source_lang = sentences.name
         target_lang = references.name
-        translator = Translator(self.__logger, model, tokenizer, source_lang, target_lang)
 
-        translated_sentences = sentences.map(lambda s: translator.translate(s))
-        translated_sentences = translated_sentences.map(lambda s: translator.normalize_translation(s))
+        translated_sentences: pd.Series = sentences.map(lambda s: self.__translator.translate(s, source_lang, target_lang))
+
+        return (self.evaluate_bleu(translated_sentences, references), self.evaluate_chrf(translated_sentences, references))
```
Comment (filip-szweda) [inline model_utilities/evaluate/evaluator.py:23]:
```suggestion
        return self.evaluate_bleu(translated_sentences, references), self.evaluate_chrf(translated_sentences, references)
```

File: model_utilities/evaluate/evaluator.py [hunk 6]
```diff
@@ -10,28 +10,21 @@
...
 
-        translated_sentences, references = self.__remove_corrupted_sentences(translated_sentences, references)
+    def evaluate_bleu(self, sentences: pd.Series, references: pd.Series) -> sacrebleu.metrics.BLEUScore:
```
Comment (filip-szweda) [inline model_utilities/evaluate/evaluator.py:25]: Unnecessary function, it only transforms the arguments and passes them forward.

File: model_utilities/translate/translator.py [hunk 7]
```diff
@@ -1,51 +1,36 @@
...
-        )
-        return self.__tokenizer.batch_decode(result)
+        self.__max_length = max_length
+
+        self.__pipeline = pipeline("translation",
+                                   model=self.__model,
+                                   tokenizer=self.__tokenizer,
+                                   max_length=self.__max_length)
+
+    # def normalize_translation(self, input: list[str]) -> str | None:
```
Comment (filip-szweda) [inline model_utilities/translate/translator.py:26]: Throw out, not needed.

File: model_utilities/__main__.py [hunk 8]
```diff
@@ -24,30 +24,38 @@ def train_model(config: dict, logger: Logger) -> None:
 def use_model(config: dict, logger: Logger) -> None:
     output_model_name = config["MODEL"]["output_model_name"]
 
-    pretrained_model = AutoModelForSeq2SeqLM.from_pretrained(output_model_name)
+    # Default model commented out to make comparing results easy
```
Comment (filip-szweda) [inline model_utilities/__main__.py:27]: For debug, create a new package command instead.

File: model_utilities/translate/translator.py [hunk 9]
```diff
@@ -1,51 +1,36 @@
...
+    #     # Technically matching against 'Latn' will break if we do languages other than PL and CSB but we won't
+    #     match = re.match(r"^.*Latn (.*)<.*$", input[0])
+    #     # Workaround for when model spits out a translation that is too long and doesn't match regex.
+    #     if match == None:
+    #         self.__logger.warning(f"Warning translation does not match regex pattern: {input[0]}")
+    #         return None
+    #     return match.group(1)
+
+    def translate(self, text: str, source_lang: str, target_lang: str) -> str:
+        return self.__pipeline(text, src_lang=source_lang, tgt_lang=target_lang)[0]["translation_text"]
```
Comment (filip-szweda) [inline model_utilities/translate/translator.py:36]:
```python
translation_root = self.__pipeline(text, src_lang=source_lang, tgt_lang=target_lang)[0]
```

File: model_utilities/__main__.py (hunk 3 above)
Comment (jan-Kulpas) [reply model_utilities/__main__.py:52]: This looks even worse and isnt that much useful

File: model_utilities/translate/translator.py [hunk 10]
```diff
@@ -1,51 +1,28 @@
...
-        result = self.__model.generate(
-            **inputs.to(self.__model.device),
-            forced_bos_token_id=self.__forced_bos_token_id,
-            max_new_tokens=int(self.__a + self.__b * inputs.input_ids.shape[1]),
-            num_beams=self.__num_beams, **kwargs
-        )
-        return self.__tokenizer.batch_decode(result)
+        self.__max_length = max_length
+
+        self.__pipeline = pipeline("translation",
```
Comment (filip-szweda) [inline model_utilities/translate/translator.py:21]:
```suggestion
        self.__pipeline = pipeline(task="translation",
```

File: model_utilities/translate/translator.py [hunk 11]
```diff
@@ -1,51 +1,28 @@
...
+                                   model=self.__model,
+                                   tokenizer=self.__tokenizer,
+                                   max_length=self.__max_length)
+
+    def translate(self, text: str, source_lang: str, target_lang: str) -> str:
```
Comment (filip-szweda) [inline model_utilities/translate/translator.py:26]: The logger is currently unused in the translator.
```python
def translate(self, text: str, source_lang: str, target_lang: str) -> str:
        model_output = self.__pipeline(text, src_lang=source_lang, tgt_lang=target_lang)
        translation_text = model_output[0]["translation_text"]
        self.__logger.info(f"Translation from {source_lang} to {target_lang}: {translation_text}")
        return translation_text
```

File: model_utilities/__main__.py [hunk 12]
```diff
@@ -24,30 +24,34 @@ def train_model(config: dict, logger: Logger) -> None:
...
 
-    pretrained_model = AutoModelForSeq2SeqLM.from_pretrained(output_model_name)
+    model = AutoModelForSeq2SeqLM.from_pretrained(output_model_name)
     tokenizer = NllbTokenizer.from_pretrained(output_model_name)
     message = "Wsiądźmy do tego autobusu"
-    translated_message = Translator(logger, pretrained_model, tokenizer).translate(message)
+
+    translated_message = Translator(logger, model, tokenizer).translate(message, "pol_Latn", "csb_Latn")
+
     print(f"Message {message} has been translated to: {translated_message}")
```
Comment (filip-szweda) [inline model_utilities/__main__.py:33]: Use the logger instead.

**Comment by filip-szweda** [comment]
Change CUDA and configuration logging level to DEBUG.

**Pull request merged**